### PR TITLE
Ewb 1075 use asyncio grpc

### PR DESCRIPTION
**Pull request by Esaotuelr**

### CODE REVIEW

**Comment by ryanjcoleman** [comment]
Task linked: [EWB-1075 Utilise Python gRPC Asyncio API](https://app.clickup.com/t/6929263/EWB-1075)

File: src/zepben/evolve/streaming/grpc/connect.py [hunk 1]
```diff
@@ -9,6 +9,7 @@
 #  file, You can obtain one at https://mozilla.org/MPL/2.0/.
 import contextlib
 import warnings
+from asyncio import get_event_loop
```
Comment (charlta) [inline src/zepben/evolve/streaming/grpc/connect.py:12]: Remove import that was left from debugging

**Review by charlta — CHANGES_REQUESTED** [review]
You will need to deploy a new evolve-grpc that updates the conflicting depnedency

Can you also please add an entry to the change log for your update

File: changelog.md [hunk 2]
```diff
@@ -24,10 +24,12 @@
   * `function: TransformerFunctionKind`
 
 ##### Enhancements
-* None.
+* Enable asyncio for gRPC from the newest update of grpcio and grpcio-tool
 
 ##### Fixes
 * Updated gRPC to fix support for latest LetsEncrypt certificates.
+* By pass grpc.channel_ready_future() for async io path.
+* Ensure insecure channel is used on async io path when secure is set to False.
```
Comment (charlta) [inline changelog.md:32]: No need for these two lines as they are fixes for the current feature 'Enable asyncio for gRPC from the newest update of grpcio and grpcio-tool'.

**Review by charlta — CHANGES_REQUESTED** [review]

**Review by kgreav — APPROVED** [review]

**Review by charlta — APPROVED** [review]

**Comment by charlta** [comment]
@kgreav @Esaotuelr Just need to fix whatever tests are failing

**Comment by kgreav** [comment]
So tests are failing because of the lack of asyncio support in grpc_testing. From what I can find they have no near term intention to support it (they haven't even roadmapped it). I think it's reasonably possible to add support direct to grpc_testing ourselves and we should probably do that before merging this PR.

The alternative seems to be spinning up an actual server, which looks like it will be more effort.

**Comment by kgreav** [comment]
So I think the async changes also seem to break the SyncNetworkConsumerClient. When you use it and do a get_network_hierarchy() I get an error. @Lax125 can you test this?

**Comment by Lax125** [comment]
Works on my machine™. Output for `run_streaming.py`:
```
Running streaming...
('32bit', 'WindowsPE')
   Running get object...
      Exception caught: ValueError - No object with mRID 21527151-6fce-423d-84e5-8254a00b05b1 could be found.
   get object took perf=0:00, proc=0:00
   Running get feeder...
      Exception caught: ValueError - No object with mRID CTN005 could be found.
   get feeder took perf=0:00, proc=0:00
   Running retrieve network hierarchy...
      Num geographical regions: 1
      Num sub geographical regions: 1
      Num substations: 1
      Num feeders: 3
      Num circuits: 0
      Num loops: 0
   retrieve network hierarchy took perf=0:00, proc=0:00
streaming took perf=0:00, proc=0:00

Process finished with exit code 0
```
...which is exactly the same as on the main branch. Show me the error you get.

**Comment by charlta** [comment]
You need to set the mRID's in `run_streaming.py` to match those in the database you have loaded, otherwise this is expected behaviour.

If the mRID's are valid, then you have something to look at

**Comment by Lax125** [comment]
I know, those are the first two parts. It retrieved the network hierarchy just fine in the last part, which is what I was pointing out.

**Comment by kgreav** [comment]
i'll have to play around, must have been something in my environment.

File: src/zepben/evolve/streaming/get/async_utils.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+from concurrent.futures import ThreadPoolExecutor
+from typing import Callable, TypeVar, Awaitable
+
+__all__ = ["make_async"]
+
+
+T = TypeVar('T')
+
+
+def make_async(func: Callable[..., T], timeout=None) -> Callable[..., Awaitable[T]]:
```
Comment (kgreav) [inline src/zepben/evolve/streaming/get/async_utils.py:17]: This breaks pretty horribly. I don't know why it doesn't break in tests - something we need to look into.

```
Traceback (most recent call last):
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/common.pyx.pxi", line 182, in grpc._cython.cygrpc.get_working_loop
RuntimeError: no running event loop

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/krut/work/git/cim-to-opendss-processor/src/zepben/cim2opendss/main.py", line 324, in <module>
    start()
  File "/home/krut/work/git/cim-to-opendss-processor/src/zepben/cim2opendss/main.py", line 320, in start
    asyncio.run(main())
  File "/usr/lib64/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib64/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/home/krut/work/git/cim-to-opendss-processor/src/zepben/cim2opendss/main.py", line 272, in main
    network_hierarchy = (await client.get_network_hierarchy()).throw_on_error().value
  File "/home/krut/work/git/cimbend/src/zepben/evolve/streaming/grpc/grpc.py", line 78, in throw_on_error
    raise self.result
  File "/home/krut/work/git/cimbend/src/zepben/evolve/streaming/grpc/grpc.py", line 106, in try_rpc
    return GrpcResult(await rpc())
  File "/home/krut/work/git/cimbend/src/zepben/evolve/streaming/get/network_consumer.py", line 413, in _handle_network_hierarchy
    response = await make_async(self._stub.getNetworkHierarchy)(GetNetworkHierarchyRequest())
  File "/home/krut/work/git/cimbend/src/zepben/evolve/streaming/get/async_utils.py", line 28, in inner
    return await asyncio.wait_for(loop.run_in_executor(executor, functools.partial(func, *args, **kwargs)), timeout=timeout)
  File "/usr/lib64/python3.9/asyncio/tasks.py", line 442, in wait_for
    return await fut
  File "/usr/lib64/python3.9/concurrent/futures/thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/krut/work/git/cim-to-opendss-processor/venv/lib64/python3.9/site-packages/grpc/aio/_channel.py", line 132, in __call__
    call = UnaryUnaryCall(request, _timeout_to_deadline(timeout),
  File "/home/krut/work/git/cim-to-opendss-processor/venv/lib64/python3.9/site-packages/grpc/aio/_call.py", line 503, in __init__
    channel.call(method, deadline, credentials, wait_for_ready),
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/channel.pyx.pxi", line 134, in grpc._cython.cygrpc.AioChannel.call
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi", line 49, in grpc._cython.cygrpc._AioCall.__cinit__
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/grpc_aio.pyx.pxi", line 97, in grpc._cython.cygrpc.init_grpc_aio
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/grpc_aio.pyx.pxi", line 101, in grpc._cython.cygrpc.init_grpc_aio
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/grpc_aio.pyx.pxi", line 86, in grpc._cython.cygrpc._initialize_per_loop
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/common.pyx.pxi", line 184, in grpc._cython.cygrpc.get_working_loop
  File "/usr/lib64/python3.9/asyncio/events.py", line 642, in get_event_loop
    raise RuntimeError('There is no current event loop in thread %r.'
RuntimeError: There is no current event loop in thread 'ThreadPoolExecutor-0_0'.
```

Comment (charlta) [reply src/zepben/evolve/streaming/get/async_utils.py]: No longer an issue as this has been removed

**Review by charlta — CHANGES_REQUESTED** [review]
@Lax125 as discussed, can you please put any attribution for the gRPC testing code

**Comment by Lax125** [comment]
The original code was licensed under See https://www.apache.org/licenses/LICENSE-2.0 Section 4 for details on copyright for modifications.

**Review by charlta — APPROVED** [review]

**Pull request merged**